### PR TITLE
Parital swagger API documentation

### LINKED ISSUE #35: [Request] API documentation with Swagger
**Is your feature request related to a problem? Please describe.**
I want to implement a client for the API and would like an easy accessible documentation for the endpoints and its responses.
Unfortunately the current jsdoc seems more aimed towards developers of the  G5API application and not consumers of the API.

**Describe the solution you'd like**
This can be done with Swagger as mentioned in another issue. There are several ways to realize this but I think the most prudent would be to have the swagger definitions close to the implementation as possible so it easier to keep it up to date.

The package [swagger-express](https://www.npmjs.com/package/swagger-express) can get the config from different places where jsdoc is one of them.
I'm not sure if swagger-express does this but swagger supports making request directly from swagger UI and can also show examples of the expected response for the endpoints.

Example of swagger ui for G5API:
image
Example of detailed information for an endpoint:
image

See here for an full example of an API documented with swagger UI: https://petstore.swagger.io/

**Describe alternatives you've considered**
I'm trying out Haskell and a project called Servant that can generate code for clients in different languages as well as swagger documentation, but this is duplication of work and it is likely that this will diverge from the real implementation.

There are also a lot of other documentation projects that can be used but one of the big pros with swagger is that it can be generated and would not need that much manual upkeep.

**Additional context**
I might implement swagger-express to see how it works and create a PR.

### PR DESCRIPTION
**Pull request by ringvold**
Swagger implemented for users, leaderboard and mapstats endpoints.
Documentation available on path `/api-docs`.

I planned to keep the original jsdoc comments but swagger did not 
work without it so they had to go. But does it not look like the jsdocs 
on the express routes was picked up by jsdocs anyways.

Changed some routes/paths to be more "RESTy". Create, update and
delete routes could go straight on root path for the resources.
Ex: `POST /users` instead of `POST /users/create`.
I can undo the route/path changes if you don't approve of it. :)

Resolves #35.

### CODE REVIEW

**Comment by PhlexPlexico** [comment]
Hmm... I'm not quite sure on the changing endpoints bit. I was mainly using the old end-points so that anyone who used get5-web would be able to easily reference and create new pages based on that. _However_ since this an API and it is not showing any output and those types of endpoints can be left on the front-end anyways. I'm going to lean more towards yes to changing them for now, but I may need some more convincing. I feel like it is a step in the right direction.

That being said, the tests will have to be updated as well then, to ensure that Github doesn't yell at someone each time they make a change. I know it's probably going to yell each time there's a steam API call as well, but those are easy to look at and may be another issue to spin off for a later date.

File: app.js [hunk 1]
```diff
@@ -74,6 +80,23 @@ app.use(cors());
...
+      title: 'G5API', // Title (required)
+      version: '0.1.0', // Version (required)
+    },
+  },
+  // Path to the API docs
+  apis: ['./routes/leaderboard.js','./routes/mapstats.js','./routes/users.js'],
+};
+const swaggerSpec = swaggerJSDoc(options);
+
+app.use('/api-docs', swaggerUi.serve, swaggerUi.setup(swaggerSpec));
```
Comment (PhlexPlexico) [inline app.js:98]: Maybe we should consider a different end point like `/docs/api` instead? I'm unsure, what are your thoughts?

File: routes/mapstats.js [hunk 2]
```diff
@@ -66,20 +124,48 @@ router.get("/:match_id", async (req, res, next) => {
...
+ *         $ref: '#/components/responses/MatchAlreadyFinished'
+ *       500:
+ *         $ref: '#/components/responses/Error'
  */
-router.post("/create", Utils.ensureAuthenticated, async (req, res, next) => {
+router.post("/", Utils.ensureAuthenticated, async (req, res, next) => {
   try {
     if (req.body[0].match_id == null) {
-      res.status(404).json({ message: "Match ID Not Provided" });
+      res.status(400).json({ message: "Match ID Not Provided" });
```
Comment (PhlexPlexico) [inline routes/mapstats.js:168]: Make sure to change the tests as well if you're changing out some of the response codes, the tests under `__test__` do look for some specific status responses to ensure we hit edge cases.

File: routes/mapstats.js [hunk 3]
```diff
@@ -93,15 +179,15 @@ router.post("/create", Utils.ensureAuthenticated, async (req, res, next) => {
       !Utils.superAdminCheck(req.user)
     ) {
       res
-        .status(401)
+        .status(403)
```
Comment (PhlexPlexico) [inline routes/mapstats.js:182]: Same as above.

File: routes/mapstats.js [hunk 4]
```diff
@@ -93,15 +179,15 @@ router.post("/create", Utils.ensureAuthenticated, async (req, res, next) => {
...
         .json({ message: "User is not authorized to perform action." });
       return;
     } else if (
       matchRow[0].cancelled == 1 ||
       matchRow[0].forfeit == 1 ||
       matchRow[0].end_time != null
     ) {
-      res.status(401).json({ message: "Match is already finished." });
+      res.status(422).json({ message: "Match is already finished." });
```
Comment (PhlexPlexico) [inline routes/mapstats.js:190]: Same as above.

File: routes/users.js [hunk 5]
```diff
@@ -87,27 +198,54 @@ router.post("/create", Utils.ensureAuthenticated, async (req, res, next) => {
         res.json({ message: "User created successfully" });
       });
     } else {
-      res.status(401).json({ message: "You are not authorized to do this." });
+      res.status(403).json({ message: "You are not authorized to do this." });
```
Comment (PhlexPlexico) [inline routes/users.js:201]: Just make sure the tests are adjusted :)

File: routes/users.js [hunk 6]
```diff
@@ -87,27 +198,54 @@ router.post("/create", Utils.ensureAuthenticated, async (req, res, next) => {
...
+ *         description: Nothing to update
+ *         content:
+ *           application/json:
+ *             schema:
+ *               $ref: '#/components/schemas/SimpleResponse'
+ *       500:
+ *         $ref: '#/components/responses/Error'
  */
-router.put("/update", Utils.ensureAuthenticated, async (req, res, next) => {
+router.put("/", Utils.ensureAuthenticated, async (req, res, next) => {
```
Comment (PhlexPlexico) [inline routes/users.js:248]: As I've stated in the PR, I'm not opposed to this, this was me just importing what was directly from get5-web. I'll need to sleep on it for now, but don't remove these just yet! And once again on a comment, just gotta update test cases. I think I can even give it a go and update them as needed as well.

**Comment by ringvold** [comment]
I totally forgot about the tests! 
I also had a link for an article about 401 vs 403 status code that I had to check and wanted to share: https://stackoverflow.com/questions/3297048/403-forbidden-vs-401-unauthorized-http-responses
I'm used to have the POST/PUT/DELETE right on the root mostly because it was the default in rails (and probably other frameworks) so in now makes sense for me to organize CRUD operations like that.

**Comment by PhlexPlexico** [comment]
At the end of the day it's probably more organized having it as you set it as well, so feel free to keep the changes and I'll merge 'em in once everything is complete!

As for the SO link, thanks! It does make a bit more sense that it is "not authorized by design" vs "something went wrong on client side, try again" is a bit more suiting.

File: routes/users.js (hunk 5 above)
Comment (PhlexPlexico) [reply routes/users.js:201]: Ran this locally and it seems users is properly picking things up in the tests!

File: routes/mapstats.js (hunk 4 above)
Comment (PhlexPlexico) [reply routes/mapstats.js:190]: Some tests in mapstats are still failing. I think there's some legacy stuff leftover from the `update` and `create` I think!

**Comment by ringvold** [comment]
It seems like some of the test do not pass for me on master. Gameserver and mapstats does not pass on master. Does all of them pass for you on master?

**Comment by PhlexPlexico** [comment]
Very odd! Yes, I believe last I checked on `master` they were running just fine. Let me go give it another test. Just make sure you have the Steam API Key set in your `test.json` in the `config` folder! I think without the Steam Resolver, some of the user tests and game server tests fail.

File: routes/mapstats.js (hunk 4 above)
Comment (PhlexPlexico) [reply routes/mapstats.js:190]: Made a small PR to update the test cases in mapstats. Ran the tests on my end and they seem to be working now. The issue with Travis is that there are two secrets (database key, and Steam API key) so the merging tests will always fail... I need to think of a way to mock steam resolver requests. But if you merge in those changes we should be able to merge in this branch whenever you're ready!

**Comment by PhlexPlexico** [comment]
Closing in favour of #42. Thanks for the help, and I'll continue getting Swagger in there going forward :)
